src/app/rosetta/lib/block.ml

```diff
@@ -0,0 +1,579 @@
+
+  module T (M : Monad_fail.S) = struct
+    let of_partial_identifier (identifier : Partial_block_identifier.t) =
+      match (identifier.index, identifier.hash) with
+      | None, None ->
+          M.fail
+            (Errors.create
+               ~context:
+                 "Partial block identifier must have at least an index or a \
+                  hash, it cannot be empty"
+               (`Json_parse None))
+      | Some index, None ->
+          M.return (`This (`Height index))
+      | None, Some hash ->
+          M.return (`That (`Hash hash))
+      | Some index, Some hash ->
+          M.return (`Those (`Height index, `Hash hash))
+  end
+end
+
+module Op = struct
+  type 'a t = {label: 'a; related_to: 'a option} [@@deriving eq]
+
+  let build ~a_eq ~plan ~f =
+    List.mapi plan ~f:(fun i op ->
+        let operation_identifier i =
+          {Operation_identifier.index= Int64.of_int_exn i; network_index= None}
+        in
+        let related_operations =
+          match op.related_to with
+          | Some relate ->
+              List.findi plan ~f:(fun _ a -> a_eq relate a.label)
+              |> Option.map ~f:(fun (i, _) -> operation_identifier i)
+              |> Option.to_list
+          | None ->
+              []
+        in
+        f ~related_operations ~operation_identifier:(operation_identifier i) op
+    )
+end
+
+(* TODO: Populate postgres DB with at least one of each kind of transaction and then make sure ops make sense *)
+module User_command_info = struct
+  module Kind = struct
+    type t =
+      [`Payment | `Delegation | `Create_token | `Create_account | `Mint_tokens]
+  end
+
+  module Account_creation_fees_paid = struct
+    type t =
+      | By_no_one
+      | By_fee_payer of Unsigned.UInt64.t
+      | By_receiver of Unsigned.UInt64.t
+    [@@deriving eq]
+  end
+
+  module Failure_status = struct
+    type t = [`Applied of Account_creation_fees_paid.t | `Failed of string]
+    [@@deriving eq]
+  end
+
+  type t =
+    { kind: Kind.t
+    ; fee_payer: [`Pk of string]
+    ; source: [`Pk of string]
+    ; receiver: [`Pk of string]
+    ; fee_token: Unsigned.UInt64.t
+    ; token: Unsigned.UInt64.t
+    ; fee: Unsigned.UInt64.t
+    ; nonce: Unsigned.UInt32.t
+    ; amount: Unsigned.UInt64.t option
+    ; hash: string
+    ; failure_status: Failure_status.t option }
+
+  let to_operations (t : t) : Operation.t list =
+    (* First build a plan. The plan specifies all operations ahead of time so
+     * we can later compute indices and relations when we're building the full
+     * models.
+     *
+     * For now, relations will be defined only on the two sides of a given
+     * transfer. ie. Source decreases, and receiver increases.
+    *)
+    let plan : 'a Op.t list =
+      (* The dec side of a user command's fee transfer is here *)
+      (* TODO: Relate fee_payer_dec with fee_receiver_inc across user_commands and internal commands *)
+      [{Op.label= `Fee_payer_dec; related_to= None}]
```

Comment:
This should be added only if fee is non-zero. If say, the total fee within a block is zero then you would not have a `Fee_receiver_inc` op to cancel it out